### ISSUE TITLE
About the batch dim and get_last_hidden

### ISSUE THREAD

**Issue by takuyara** [hobincar/SA-LSTM#15]
Hi hobincar, I see that there's some difference in nn.Linear and nn.LSTM when it comes to the batch dim. nn.Linear is batch_first while nn.LSTM is batch_middle. How did you manage that?
The get_last_hidden function in Decoder class is quite confusing for me. Can you demonstrate those complicated tensor operations for me? What does this function do? Many thanks.

**Comment by hobincar** [comment]
Hi, takuyara.

Q.  I see that there's some difference in nn.Linear and nn.LSTM when it comes to the batch dim. nn.Linear is batch_first while nn.LSTM is batch_middle. How did you manage that?
A. I make the input-to-LSTM to be batch-middle by [this line](https://github.com/hobincar/SA-LSTM/blob/master/models/caption_generator.py#L31), and then the input-to-Linear to be batch-first by [this line](https://github.com/hobincar/RecNet/blob/master/models/decoder.py#L59). The latter is possible because we feed word by word to LSTM, which makes the length of LSTM's output to be 1.

Q. The get_last_hidden function in Decoder class is quite confusing for me. Can you demonstrate those complicated tensor operations for me? What does this function do?
A. It is for the temporal attention module which takes the last hidden state as input. It is complex since I tried to implement the function `get_last_hidden` to work regardless of the types of RNN (GRU/LSTM), the number of layers, and the number of directions. What it does is literally just get the last hidden state.
